storage/mutable_forest.go

```diff
@@ -0,0 +1,159 @@
+package storage
+
+import (
+	"fmt"
+
+	dbm "github.com/tendermint/tendermint/libs/db"
+)
+
+const (
+	commitsPrefix = "c"
+	treePrefix    = "t"
+)
+
+type ForestReader interface {
+	Reader(prefix []byte) (KVCallbackIterableReader, error)
+}
+
+// MutableForest is a collection of versioned lazily-loaded RWTrees organised by prefix. It maintains a global state hash
+// by storing CommitIDs in a special commitsTree (you could think of it is a two-layer single tree rather than a forest).
+//
+// The trees (or sub-trees if you prefer) in the forest are RWTrees which wrap an IAVL MutableTree routing writes to the
+// MutableTree and reads to the last saved ImmutableTree. In this way reads act only against committed state and can be
+// lock free (this allows us to avoid blocking commits - particularly for long-running iterations).
+//
+// The trees in the forest are created lazily as required by new writes. There is a cache of most recently used trees
+// and trees that may require a save are marked as such. New writes are only available to read after a Save().
+//
+// Here is an example forest (the output is generated by the Dump() function):
+//  .
+//  ├── balances
+//  │   ├── Caitlin -> 2344
+//  │   ├── Cora -> 654456
+//  │   ├── Edward -> 34
+//  │   └── Lindsay -> 654
+//  └── names
+//      ├── Caitlin -> female
+//      ├── Cora -> female
+//      ├── Edward -> male
+//      └── Lindsay -> unisex
+//
+// Here there are two tree indexed by the prefixes 'balances' and 'names'.
+//
+// To perform reads of the forest we access it in the following way:
+//
+//   tree, err := forest.Reader("names")
+//   gender := tree.Get("Cora")
+//
+// To perform writes:
+//
+//   tree, err := forest.Writer("names")
+//   tree.Set("Cora", "unspecified")
+//
+// If there is no tree currently stored at the prefix passed then it will be created when the forest is saved:
+//
+//  hash, version, err := forest.Save()
+//
+// where the global version for the forest is returned.
+
+type MutableForest struct {
```

Comment:
Great description and implementation

Reply:
Glad you agree... went through a number of indecisions...